domain/secret/state/state.go

```diff
@@ -1884,7 +1889,151 @@ ON CONFLICT(secret_id) DO UPDATE SET
 		if err := tx.Query(ctx, insertLatestStmt, secret).Run(); err != nil {
 			return errors.Annotatef(err, "updating latest revision %d for cross model secret %q", latestRevision, uri)
 		}
+		if err := markObsoleteRevisions(ctx, st, tx, uri); err != nil {
+			return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+		}
 		return nil
 	})
 	return errors.Trace(domain.CoerceError(err))
 }
+
+// InitialWatchStatementForObsolete returns the initial watch statement and the table name for watching obsolete revisions.
+func (st State) InitialWatchStatementForObsoleteRevision(
+	ctx context.Context, appOwners domainsecret.ApplicationOwners, unitOwners domainsecret.UnitOwners,
+) (string, string) {
+	tableName := "secret_revision"
+
+	q := `
+SELECT uuid
+FROM secret_revision sr
+`[1:]
+
+	// The condition is empty because we want to watch all obsolete revisions.
+	if len(appOwners) == 0 && len(unitOwners) == 0 {
+		return tableName, q
+	}
+
+	var join, condition string
+	if len(unitOwners) > 0 && len(appOwners) > 0 {
+		join = `
+JOIN secret_application_owner sao ON sr.secret_id = sao.secret_id
```

Comment:
This can be made easier to read using WITH statements. See listCharmSecrets. That also has the use of sqlair IN

Reply:
I'd think simple JOIN is way better than WITH?